### PR TITLE
Prettify the export format of NAS trainer

### PR DESCRIPTION
**Pull request by ultmaster**
Pending #2386 as it uses APIs introduced in that PR. I coded on another branch merging #2386 and cherry picked the changes.

**Note that this PR will be backward-compatible itself. Old checkpoints will still be valid.**

This PR prettifies the export format of NAS trainer (point 2 of #2316). Here are three examples (of the new export format):

P-DARTS (first):

```
{
  "normal_n2_p0": "maxpool",
  "normal_n2_p1": "maxpool",
  "normal_n2_switch": [
    "normal_n2_p0",
    "normal_n2_p1"
  ],
  "normal_n3_p0": [],
  "normal_n3_p1": "maxpool",
  "normal_n3_p2": "maxpool",
  "normal_n3_switch": [
    "normal_n3_p1",
    "normal_n3_p2"
  ],
  "reduce_n2_p0": "maxpool",
  "reduce_n2_p1": "dilconv5x5",
  "reduce_n2_switch": [
    "reduce_n2_p0",
    "reduce_n2_p1"
  ],
  "reduce_n3_p0": "sepconv5x5",
  "reduce_n3_p1": "maxpool",
  "reduce_n3_p2": [],
  "reduce_n3_switch": [
    "reduce_n3_p0",
    "reduce_n3_p1"
  ]
}
```

ENAS macro:

```
{
  "InputChoice11": [
    "layer_2",
    "layer_3"
  ],
  "InputChoice13": "layer_4",
  "InputChoice15": [
    "layer_2",
    "layer_3"
  ],
  "InputChoice17": [
    "layer_2",
    "layer_4"
  ],
  "InputChoice19": [
    "layer_0",
    "layer_2",
    "layer_8"
  ],
  "InputChoice21": [
    "layer_0",
    "layer_1",
    "layer_7"
  ],
  "InputChoice23": [
    "layer_4",
    "layer_5",
    "layer_9"
  ],
  "InputChoice3": "layer_0",
  "InputChoice5": "layer_0",
  "InputChoice7": [],
  "InputChoice9": "layer_0",
  "LayerChoice1": 2,
  "LayerChoice10": 3,
  "LayerChoice12": 2,
  "LayerChoice14": 2,
  "LayerChoice16": 0,
  "LayerChoice18": 1,
  "LayerChoice2": 4,
  "LayerChoice20": 4,
  "LayerChoice22": 4,
  "LayerChoice4": 4,
  "LayerChoice6": 3,
  "LayerChoice8": 1
}
```

ENAS micro:

```
{
  "normal_node_0_x_input": 0,
  "normal_node_0_x_op": 3,
  "normal_node_0_y_input": 1,
  "normal_node_0_y_op": 4,
  "normal_node_1_x_input": 2,
  "normal_node_1_x_op": 0,
  "normal_node_1_y_input": 2,
  "normal_node_1_y_op": 0,
  "normal_node_2_x_input": 3,
  "normal_node_2_x_op": 2,
  "normal_node_2_y_input": 2,
  "normal_node_2_y_op": 0,
  "normal_node_3_x_input": 4,
  "normal_node_3_x_op": 2,
  "normal_node_3_y_input": 0,
  "normal_node_3_y_op": 1,
  "normal_node_4_x_input": 5,
  "normal_node_4_x_op": 2,
  "normal_node_4_y_input": 1,
  "normal_node_4_y_op": 1,
  "reduce_node_0_x_input": 0,
  "reduce_node_0_x_op": 1,
  "reduce_node_0_y_input": 1,
  "reduce_node_0_y_op": 1,
  "reduce_node_1_x_input": 0,
  "reduce_node_1_x_op": 1,
  "reduce_node_1_y_input": 0,
  "reduce_node_1_y_op": 0,
  "reduce_node_2_x_input": 0,
  "reduce_node_2_x_op": 4,
  "reduce_node_2_y_input": 1,
  "reduce_node_2_y_op": 0,
  "reduce_node_3_x_input": 4,
  "reduce_node_3_x_op": 0,
  "reduce_node_3_y_input": 1,
  "reduce_node_3_y_op": 4,
  "reduce_node_4_x_input": 4,
  "reduce_node_4_x_op": 2,
  "reduce_node_4_y_input": 0,
  "reduce_node_4_y_op": 1
}
```

### CODE REVIEW

**Comment by QuanluZhang** [comment]
please fix pylint error

File: src/sdk/pynni/nni/nas/pytorch/fixed.py [hunk 1]
```diff
@@ -33,6 +34,32 @@ def __init__(self, model, fixed_arc, strict=True):
             raise RuntimeError("Unexpected keys found in fixed architecture: {}.".format(fixed_arc_keys - mutable_keys))
         if mutable_keys - fixed_arc_keys:
             raise RuntimeError("Missing keys in fixed architecture: {}.".format(mutable_keys - fixed_arc_keys))
+        self._fixed_arc = self._convert_human_readable_architecture(self._fixed_arc)
+
+    def _convert_human_readable_architecture(self, human_arc):
+        result_arc = {k: to_list(v) for k, v in human_arc.items()}  # there could be tensors, numpy arrays, etc.
+        # First, convert non-list to list.
+        result_arc = {k: v if isinstance(v, list) else [v] for k, v in result_arc.items()}
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/fixed.py:42]: why not combine line40 and line42 by creating a list for non list in `to_list` function?

File: src/sdk/pynni/nni/nas/pytorch/fixed.py [hunk 2]
```diff
@@ -33,6 +34,32 @@ def __init__(self, model, fixed_arc, strict=True):
...
+        # Second, infer which ones are multi-hot arrays and which ones are in human-readable format.
+        # This is non-trivial, since if an array in [0, 1], we cannot know for sure it means [false, true] or [true, true].
+        # Here, we assume an multihot array has to be a boolean array or a float array and matches the length.
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/fixed.py:45]: in what case the multihot array is float?

Comment (QuanluZhang) [reply src/sdk/pynni/nni/nas/pytorch/fixed.py:45]: so the exported final arch could be human readable or multihot?

**Comment by QuanluZhang** [comment]
please update doc accordingly

File: src/sdk/pynni/nni/nas/pytorch/fixed.py (hunk 2 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/fixed.py:45]: Yes. We can at least be backward-compatible on that.
Multi-hot array is float if it's a weighted-sum of candidates. By current design, we allow final architectures to be a mixture of conv3x3 and con5x5, for example.

File: src/sdk/pynni/nni/nas/pytorch/fixed.py (hunk 1 above)
Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/fixed.py:42]: I don't think so. `to_list` is to be convert non-python lists (e.g., numpy array or tensors) to python lists without affecting the semantics. In line 42, it changes single value to a list of one single value. If I merge it into `to_list`, `to_list` might be confusing to whoever tries to use it in their own code.

Comment (ultmaster) [reply src/sdk/pynni/nni/nas/pytorch/fixed.py:42]: btw, I've made it more clear in comments.

File: docs/en_US/NAS/NasGuide.md [hunk 3]
```diff
@@ -156,12 +156,12 @@ model = Net()
...
 
-The JSON is simply a mapping from mutable keys to one-hot or multi-hot representation of choices. For example
+The JSON is simply a mapping from mutable keys to choices. For example
 
 ```json
 {
-    "LayerChoice1": [false, true, false, false],
-    "InputChoice2": [true, true, false]
+    "LayerChoice1": "conv5x5",
+    "InputChoice2": [1, 2],
```
Comment (QuanluZhang) [inline docs/en_US/NAS/NasGuide.md:164]: what is the meaning of 1, 2? the index?

File: src/sdk/pynni/nni/nas/pytorch/fixed.py [hunk 4]
```diff
@@ -33,6 +32,33 @@ def __init__(self, model, fixed_arc, strict=True):
             raise RuntimeError("Unexpected keys found in fixed architecture: {}.".format(fixed_arc_keys - mutable_keys))
         if mutable_keys - fixed_arc_keys:
             raise RuntimeError("Missing keys in fixed architecture: {}.".format(mutable_keys - fixed_arc_keys))
+        self._fixed_arc = self._convert_human_readable_architecture(self._fixed_arc)
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/fixed.py:35]: this function is convert to or from human readable architecture?

File: src/sdk/pynni/nni/nas/pytorch/fixed.py [hunk 5]
```diff
@@ -33,6 +32,33 @@ def __init__(self, model, fixed_arc, strict=True):
...
+        # First, convert non-list to list, because there could be {"op1": 0} or {"op1": "conv"},
+        # which means {"op1": [0, ]} ir {"op1": ["conv", ]}
+        result_arc = {k: v if isinstance(v, list) else [v] for k, v in result_arc.items()}
+        # Second, infer which ones are multi-hot arrays and which ones are in human-readable format.
+        # This is non-trivial, since if an array in [0, 1], we cannot know for sure it means [false, true] or [true, true].
+        # Here, we assume an multihot array has to be a boolean array or a float array and matches the length.
+        for mutable in self.mutables:
+            if mutable.key not in result_arc:
+                continue  # skip silently
+            choice_arr = result_arc[mutable.key]
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/fixed.py:48]: what is the meaning of "arr"?

File: src/sdk/pynni/nni/nas/pytorch/fixed.py [hunk 6]
```diff
@@ -33,6 +32,33 @@ def __init__(self, model, fixed_arc, strict=True):
             raise RuntimeError("Unexpected keys found in fixed architecture: {}.".format(fixed_arc_keys - mutable_keys))
         if mutable_keys - fixed_arc_keys:
             raise RuntimeError("Missing keys in fixed architecture: {}.".format(mutable_keys - fixed_arc_keys))
+        self._fixed_arc = self._convert_human_readable_architecture(self._fixed_arc)
+
+    def _convert_human_readable_architecture(self, human_arc):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/fixed.py:37]: please provide docstring for this function, though it is private.

File: src/sdk/pynni/nni/nas/pytorch/mutator.py [hunk 7]
```diff
@@ -185,17 +196,23 @@ def on_forward_input_choice(self, mutable, tensor_list):
         mask = self._get_decision(mutable)
         assert len(mask) == mutable.n_candidates, \
             "Invalid mask, expected {} to be of length {}.".format(mask, mutable.n_candidates)
-        out = self._select_with_mask(lambda x: x, [(t,) for t in tensor_list], mask)
+        out, mask = self._select_with_mask(lambda x: x, [(t,) for t in tensor_list], mask)
         return self._tensor_reduction(mutable.reduction, out), mask
 
     def _select_with_mask(self, map_fn, candidates, mask):
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/mutator.py:202]: the new implementation of this function has complex logic, please add docstring for this function.

File: src/sdk/pynni/nni/nas/pytorch/mutator.py [hunk 8]
```diff
@@ -185,17 +196,23 @@ def on_forward_input_choice(self, mutable, tensor_list):
...
-        if "BoolTensor" in mask.type():
+        if (isinstance(mask, list) and len(mask) >= 1 and isinstance(mask[0], bool)) or \
+                (isinstance(mask, np.ndarray) and mask.dtype == np.bool) or \
+                "BoolTensor" in mask.type():
             out = [map_fn(*cand) for cand, m in zip(candidates, mask) if m]
-        elif "FloatTensor" in mask.type():
+        elif (isinstance(mask, list) and len(mask) >= 1 and isinstance(mask[0], (float, int))) or \
```
Comment (QuanluZhang) [inline src/sdk/pynni/nni/nas/pytorch/mutator.py:207]: it can be `int`?

File: docs/en_US/NAS/NasGuide.md (hunk 3 above)
Comment (QuanluZhang) [reply docs/en_US/NAS/NasGuide.md:164]: so it could be either index or name? when it is index and when it is name?

**Pull request merged**